awscli/customizations/cloudformation/modules.py

```diff
@@ -0,0 +1,525 @@
+        """
+        # print(f"find_ref {v}, props: {self.props}")
+        if v in self.props:
+            if v not in self.params:
+                # The parent tried to set a property that doesn't exist
+                # in the Parameters section of this module
+                msg = f"{v} not found in module Parameters: {self.source}"
+                raise exceptions.InvalidModuleError(msg=msg)
+            return self.props[v]
+
+        if v in self.params:
+            param = self.params[v]
+            if DEFAULT in param:
+                # Use the default value of the Parameter
+                return param[DEFAULT]
+            msg = f"{v} does not have a Default and is not a Property"
+            raise exceptions.InvalidModuleError(msg=msg)
+
+        for k in self.resources:
+            if v == k:
+                # Simply rename local references to include the module name
+                return {REF: self.name + v}
+
+        return None
+
+    # pylint: disable=too-many-branches,unused-argument
+    def resolve_sub(self, v, d, n):
+        """
+        Parse the Sub string and break it into tokens.
+
+        If we can fully resolve it, we can replace it with a string.
+
+        Use the same logic as with resolve_ref.
+        """
+        # print(f"resolve_sub v: {v}, d: {d}, n: {n}")
+        words = parse_sub(v, True)
+        sub = ""
+        need_sub = False
+        for word in words:
+            if word.t == WordType.STR:
+                sub += word.w
+            elif word.t == WordType.AWS:
+                sub += "${AWS::" + word.w + "}"
+                need_sub = True
+            elif word.t == WordType.REF:
+                resolved = "${" + word.w + "}"
+                need_sub = True
+                found = self.find_ref(word.w)
+                if found is not None:
+                    if isinstance(found, str):
+                        need_sub = False
+                        resolved = found
+                    else:
+                        if REF in found:
+                            resolved = "${" + found[REF] + "}"
+                        elif SUB in found:
+                            resolved = found[SUB]
+                sub += resolved
+            elif word.t == WordType.GETATT:
+                need_sub = True
+                tokens = word.w.split()
+                if len(tokens) != 2:
+                    msg = "GetAtt {word.w} has unexpected number of tokens"
+                    raise exceptions.InvalidModuleError(msg=msg)
+                if tokens[0] in self.resources:
+                    tokens[0] = self.name + tokens[0]
+
+        if need_sub:
+            d[n] = {SUB: sub}
+        else:
+            d[n] = sub
+
+    def resolve_getatt(self, v, d, n):
+        """
+        Resolve a GetAtt. All we do here is add the prefix.
+
+        !GetAtt Foo.Bar becomes !GetAtt ModuleNameFoo.Bar
+        """
+        if not isinstance(v, list):
```

Comment:
What about a string since GetAtt supports both

Reply:
Do you mean treating GetAtt like a Ref? Just `!GetAtt LogicalId`?

Reply:
Yea, if I'm reading this correct if GetAtt is `!GetAtt Foo.Bar` it wouldn't be a list and we would raise an error unless the yaml parser is converting it to a list.  If that's the case its also worth noting `Fn::GetAtt: "Foo.Bar"` is also valid.

this is a valid template:
```yaml
Resources:
  Bucket:
    Type: AWS::S3::Bucket
Outputs:
  BucketArn:
    Value:
      Fn::GetAtt: "Bucket.Arn"
```

Reply:
https://github.com/aws/aws-cli/blob/804662b31a62ee27491e8b879b29915ea1c17907/awscli/customizations/cloudformation/yamlhelper.py#L36-L40 While I agree with flipping `!GetAtt` into a list this will miss the conversion of `Fn::GetAtt: "foo.bar"` and those will remain as strings.

Reply:
Unless the quotes were stripped off already?